specs/derivation.md

```diff
@@ -0,0 +1,820 @@
+
+A validator that only reads from L1 (and so doesn't interact with the sequencer directly) does not need to be
+implemented in the way presented below. It does however need to derive the same blocks (i.e. it needs to be semantically
+equivalent). We do believe the architecture presented below has many advantages.
+
+## L2 Chain Derivation Pipeline
+
+[pipeline]: #l2-chain-derivation-pipeline
+
+Our architecture decomposes the derivation process into a pipeline made up of the following stages:
+
+1. L1 Traversal
+2. L1 Retrieval
+3. Channel Bank
+4. Batch Decoding (called `ChannelInReader` in the code)
+5. Payload Attributes Derivation (called `BatchQueue` in the code)
+6. Engine Queue
+
+> **TODO** can we change code names for these two things? maybe as part of a refactor
+
+The data flows flows from the start (outer) of the pipeline towards the end (inner). Each stage is able to push data to
+the next step.
+
+However, data is *processed* is reverse order. Meaning that if there is any data to be processed in the last stage, it
+will be processed first. Processing proceeds in "steps" that can be taken at each stage. We try to take as many steps as
+possible in the last (most inner) stage before taking steps any step in its outer stage, etc.
+
+This ensures that we use the data we already have before pulling more data and minimizes the latency of data traversing
+the derivation pipeline.
+
+Each stage can maintain its own inner state as necessary. In particular, each stage maintains a L1 block reference
+(number + hash) to the latest L1 block such that all data originating from previous blocks has been processed, and the
+data from that block is or has been processed.
+
+Let's briefly describe each stage of the pipeline.
+
+### L1 Traversal
+
+In the *L1 Traversal* stage, we simply read the header of the next L1 block. In normal operations, these will be new
+L1 blocks as they get created, though we can also read old blocks while syncing, or in case of an L1 [re-org][g-reorg].
+
+### L1 Retrieval
+
+In the *L1 Retrieval* stage, we read the block we get from the outer stage (L1 traversal), and extract data for it. In
+particular we extract a byte string that corresponds to the concatenation of the data in all the [batcher
+transaction][g-batcher-transaction] belonging to the block. This byte stream encodes a stream of [channel
+frames][g-channel-frame] (see the [Batch Submission Wire Format][wire-format] section for more info).
+
+This frames are parsed, then grouped them per [channel][g-channel] into a structure we call the *channel bank*.
+
+Some frames are ignored:
+
+- Frames where `frame.frame_number <= progress`, where `progress` is the number of the latest frame that was read.
+    - i.e. in case of duplicate frame, the first frame read from L1 is considered canonical.
+- Frames past the end of the channel (i.e. past the first frame marked with `frame.is_last == 1`) are ignored.
```

Comment:
"past end of the channel" → in terms of number, not in terms of order in the stream